src/vm/checker/check_db.rs

```diff
@@ -0,0 +1,113 @@
+use rusqlite::{Connection, OptionalExtension, NO_PARAMS, Row};
+use rusqlite::types::ToSql;
+
+
+use vm::types::TypeSignature;
+use vm::checker::errors::{CheckError, CheckErrors, CheckResult};
+use vm::checker::typecheck::{ContractAnalysis, FunctionType};
+
+const SQL_FAIL_MESSAGE: &str = "PANIC: SQL Failure in contract analysis.";
+
+pub struct AnalysisDatabase {
+    conn: Connection
+}
+
+impl AnalysisDatabase {
+    pub fn initialize(filename: &str) -> AnalysisDatabase {
+        let mut contract_db = AnalysisDatabase::inner_open(filename);
+        // this is the _laziest_ of structures at the moment.
+        //    more to come!
+        contract_db.execute("CREATE TABLE IF NOT EXISTS type_analysis_table
+                      (contract_identifier INTEGER PRIMARY KEY AUTOINCREMENT,
+                       contract_name TEXT NOT NULL UNIQUE,
+                       analysis TEXT NOT NULL)",
+                            NO_PARAMS);
+
+        contract_db.check_schema();
+
+        contract_db
+    }
+
+    pub fn memory() -> AnalysisDatabase {
+        AnalysisDatabase::initialize(":memory:")
+    }
+
+    pub fn open(filename: &str) -> AnalysisDatabase {
+        let contract_db = AnalysisDatabase::inner_open(filename);
+
+        contract_db.check_schema();
+        contract_db
+    }
+
+    pub fn check_schema(&self) {
+        let sql = "SELECT sql FROM sqlite_master WHERE name=?";
+        let _: String = self.conn.query_row(sql, &["type_analysis_table"],
+                                            |row| row.get(0))
+            .expect("Bad schema in analysis db initialization.");
+    }
+
+    pub fn inner_open(filename: &str) -> AnalysisDatabase {
+        let conn = Connection::open(filename)
+            .expect("Failure to open analysis db.");
+        AnalysisDatabase {
+            conn: conn
+        }
+    }
+
+    pub fn execute<P>(&mut self, sql: &str, params: P) -> usize
+    where
+        P: IntoIterator,
+        P::Item: ToSql {
+        self.conn.execute(sql, params)
+            .expect(SQL_FAIL_MESSAGE)
+    }
+
+    fn query_row<T, P, F>(&self, sql: &str, params: P, f: F) -> Option<T>
```

Comment:
This is now implemented in a generic way in `util/db.rs`, if you wanna dedup this.

Reply:
in your branch, right? If so, I'll wait until we merge.